core/chains/evm/logpoller/orm.go

```diff
@@ -400,6 +401,45 @@ func (o *DSORM) SelectUnmatchedLogIDs(ctx context.Context, limit int64) (ids []u
 	return ids, err
 }
 
+// SelectExcessLogIDs finds any logs old enough that MaxLogsKept has been exceeded for every filter they match.
+func (o *DSORM) SelectExcessLogIDs(ctx context.Context, limit int64) (rowIDs []uint64, err error) {
+	var limitClause string
+	if limit > 0 {
+		// We have to count the logs in descending order first, to know which ones to keep. But then reverse the order
+		// of them all if we want to delete only the oldest {limit} # of logs. Omitting this 2nd ORDER BY is fine if we're
+		// deleting all prunable logs, but if paging is enabled we don't want to have non-contiguous
+		// block ranges of logs in the db while waiting for the next page to be pruned
+		limitClause = fmt.Sprintf(" ORDER BY block_number, log_index LIMIT %d", limit)
+	}
+
+	// Allow SELECT query to run for up to 3 minutes. DELETE query will still have default 10s timeout
+	selectCtx, cancel := context.WithTimeout(sqlutil.WithoutDefaultTimeout(ctx), 3*time.Minute)
```

Comment:
It might be a good idea to check if we have a filter with `max_logs_kept > 0` before running the main query